### PR TITLE
Scanline beamforming

### PR DESCRIPTION
**Pull request by tristan-deep**
pixel_vs_scanline

## Summary

This PR adds native scanline reconstruction path to zea for classic focused imaging, where each transmit forms one image line. This is implemented as a subcase of pixel-based beamforming where:
- grid is one column per transmit.
- a receive apodization masking out all other transmits not corresponding to the receive scanning line.

Main additions:

- `zea.beamform.pixelgrid.scanline_receive_apodization`
- `zea.func.apply_receive_apodization`
- `zea.ops.ReceiveApodization`
- Parameters: `parameters.grid_type = "scanline"`, `parameters.enable_receive_apodization` and `parameters.flat_apodization`.

See snippet to reproduce figure below:

<details>
<summary>Snippet</summary>

```python
import os

os.environ.setdefault("KERAS_BACKEND", "jax")

import keras
import matplotlib.pyplot as plt
import numpy as np

import zea
from zea.ops import (
    ApplyWindow,
    Beamform,
    Cast,
    Demodulate,
    EnvelopeDetect,
    LogCompress,
    Normalize,
    Pipeline,
)

zea.init_device()

# Scanline (line-by-line) imaging is just the regular pixel-based DAS pipeline
# with `enable_scanline=True` (one pixel column per transmit) plus a receive
# apodization mask (`enable_receive_apodization=True`) that keeps only each
# pixel's owning transmit.

path = "hf://zeahub/zea-carotid-2023"
config = zea.Config.from_path("hf://zeahub/zea-carotid-2023/config.yaml")

with zea.Dataset(path, lazy=True) as dataset:
    file = dataset[2]
    parameters = file.load_parameters()
    raw = file.data.raw_data[:1]

parameters.update(config.parameters)
parameters.dynamic_range = (-60, 0)
parameters.grid_size_x = 400
parameters.grid_size_z = 600
raw = raw[:, parameters.selected_transmits]

# Regular pixel-based beamforming: all transmits compounded onto a shared grid.
pixel_pipeline = Pipeline(
    operations=[
        Cast(dtype="float32"),
        ApplyWindow(),
        Demodulate(),
        Beamform(enable_pfield=True, num_patches=200),
        EnvelopeDetect(),
        Normalize(),
        LogCompress(),
    ]
)
pixel_outputs = pixel_pipeline(
    **{pixel_pipeline.key: raw, **pixel_pipeline.prepare_parameters(parameters)}
)

# Scanline beamforming: same pipeline, but one pixel column per transmit
# (`enable_scanline`) with a receive apodization mask (`enable_receive_apodization`).
scanline_parameters = parameters.copy()
scanline_parameters.enable_scanline = True

scanline_pipeline = Pipeline(
    operations=[
        Cast(dtype="float32"),
        ApplyWindow(),
        Demodulate(),
        Beamform(enable_receive_apodization=True, num_patches=200),
        EnvelopeDetect(),
        Normalize(),
        LogCompress(),
    ]
)
scanline_outputs = scanline_pipeline(
    **{
        scanline_pipeline.key: raw,
        **scanline_pipeline.prepare_parameters(scanline_parameters),
    }
)

pixel_image = keras.ops.squeeze(pixel_outputs[pixel_pipeline.output_key])
scanline_image = keras.ops.squeeze(scanline_outputs[scanline_pipeline.output_key])

pixel_8bit = zea.display.to_8bit(pixel_image, dynamic_range=parameters.dynamic_range)
scanline_8bit = zea.display.to_8bit(scanline_image, dynamic_range=parameters.dynamic_range)

zea.visualize.set_mpl_style()
extent = np.asarray(parameters.extent_imshow) * 1e3
fig, axes = plt.subplots(1, 2, figsize=(11, 5), constrained_layout=True)
for ax, image, title in zip(
    axes,
    [pixel_8bit, scanline_8bit],
    ["Pixel (DAS compounding)", "Scanline (line-by-line)"],
):
    ax.imshow(image, extent=extent, cmap="gray", aspect="equal")
    ax.set_title(title)
    ax.set_xlabel("x [mm]")
    ax.set_ylabel("z [mm]")

fig.savefig("pixel_vs_scanline.png", dpi=150, bbox_inches="tight")
plt.close(fig)
print("Saved pixel_vs_scanline.png")

```

</details>

## Summary by CodeRabbit

* **New Features**
  * Added scanline beamforming mode (`enable_scanline`) with cartesian and polar scanline geometry.
  * Added optional apodization stages: aligned transmit (scanline reconstruction) and receive-apodization, controllable via pipeline flags.
* **Bug Fixes**
  * Improved robustness for scanline grid edge cases to avoid invalid numeric outputs.
  * Enforced correct configuration rules between p-field weighting and aligned/receive apodization.
* **Tests**
  * Expanded coverage for scanline grid geometry, aligned/receive apodization masking, and end-to-end equivalence.

### CODE REVIEW

File: zea/beamform/pixelgrid.py [hunk 1]
```diff
@@ -192,3 +192,91 @@ def polar_pixel_grid(
...
+    focus_distances,
+    polar_angles,
+    zlims,
+    num_depth_pixels,
+    azimuth_angles=None,
+    sector=False,
+):
+    """Pixel grid for scanline beamforming: one column of pixels per transmit.
+
+    Scanline (line-by-line) imaging is the special case of pixel-based DAS
```
Comment (wesselvannierop) [inline zea/beamform/pixelgrid.py]: Rephrase for clarity: "Scanline (line-by-line) imaging is a special case of pixel-based DAS beamforming where each transmit is beamformed to a single column of pixels."

File: zea/beamform/pixelgrid.py [hunk 2]
```diff
@@ -192,3 +192,91 @@ def polar_pixel_grid(
...
+    beamforming where each pixel is only ever insonified by a single transmit —
+    its own beam line. This builds that grid in the same
+    ``(grid_size_z, grid_size_x, 3)`` layout as :func:`cartesian_pixel_grid`
+    (one column ``n`` per transmit ``n``, ``num_depth_pixels`` rows), so it can
+    be beamformed by the regular pixel-based :class:`~zea.ops.Beamform`
+    pipeline. Pair with :func:`scanline_receive_apodization` (fed to
+    :class:`~zea.ops.ReceiveApodization`) to mask out every transmit but the
+    one that owns each pixel.
```
Comment (wesselvannierop) [inline zea/beamform/pixelgrid.py]: Also needs rephrasing: "to mask out every transmit but the line corresponding to each transmit" -- or something along those lines.

File: zea/beamform/pixelgrid.py [hunk 3]
```diff
@@ -192,3 +192,91 @@ def polar_pixel_grid(
 
     # In case of rounding errors, trim the grid to the correct number of radial pixels
     return grid[:num_radial_pixels, :, :]
+
+
+def scanline_pixel_grid(
```
Comment (wesselvannierop) [inline zea/beamform/pixelgrid.py:197]: I'm currently having a look if the logic in this function couldn't be shared with the other grid functions.

File: zea/func/ultrasound.py [hunk 4]
```diff
@@ -564,6 +565,36 @@ def envelope_detect(data, axis=-3):
...
 
 
+def apply_receive_apodization(data, apodization, with_batch_dim):
+    """Multiply aligned data by a per-pixel, per-transmit weight.
+
+    Used by :class:`~zea.ops.PfieldWeighting` and
+    :class:`~zea.ops.ReceiveApodization` to scale or mask out a transmit's
+    contribution to a pixel before the receive-aperture and transmit sums
+    (e.g. compounding, or scanline/line-by-line imaging where only one
+    transmit should contribute per pixel).
```
Comment (wesselvannierop) [inline zea/func/ultrasound.py]: Nice refactor!

File: zea/ops/pipeline.py [hunk 5]
```diff
@@ -1120,11 +1147,15 @@ def __init__(self, beamformer="delay_and_sum", num_patches=100, enable_pfield=Fa
...
             enable_pfield (bool): Whether to include pressure field weighting in the beamforming.
+            enable_receive_apodization (bool): Whether to include an explicit per-pixel,
+                per-transmit receive apodization mask (``parameters.flat_apodization``) in the
+                beamforming, e.g. to reconstruct scanline imaging (see class docstring).
         """
 
         self.beamformer_type = beamformer
         self.num_patches = num_patches
         self.enable_pfield = enable_pfield
+        self.enable_receive_apodization = enable_receive_apodization
```
Comment (wesselvannierop) [inline zea/ops/pipeline.py:1198]: We might want to XOR these

File: zea/parameters.py [hunk 6]
```diff
@@ -315,6 +330,12 @@ def distance_to_apex(self):
         "is_3d",
         "polar_limits",
         "distance_to_apex",
+        "transmit_origins",
+        "focus_distances",
+        "polar_angles",
+        "azimuth_angles",
+        "num_scanline_pixels",
+        "scanline_sector",
```
Comment (wesselvannierop) [inline zea/parameters.py]: It is a shame the grid will now be rebuild every time you do `p.set_transmits()`, but maybe there is no other way

File: zea/parameters.py [hunk 7]
```diff
@@ -231,6 +236,14 @@ class Parameters(BaseParameters):
     """Extra keyword arguments for the pressure-field computation. See
     :func:`zea.beamform.pfield.compute_pfield`. Defaults to ``{}``."""
 
+    num_scanline_pixels: int
+    """Number of depth samples per transmit line for scanline beamforming.
+    Defaults to 700."""
```
Comment (wesselvannierop) [inline zea/parameters.py]: Why not just use `grid_size_z`?

Comment (tristan-deep) [reply zea/parameters.py]: Relic from old implementation, will revert back!

File: zea/parameters.py (hunk 6 above)
Comment (tristan-deep) [reply zea/parameters.py]: Noticed this too, maybe some refactor can help... But from a practical sense doesn't seem to bad right now.

File: zea/parameters.py [hunk 8]
```diff
@@ -250,6 +263,8 @@ class Parameters(BaseParameters):
         "pixels_per_wavelength": {"dtype": np.int32, "default": 4},
         "pfield_kwargs": {"dtype": dict, "default": {}},
         "apply_lens_correction": {"dtype": bool, "default": False},  # native dtype on purpose
+        "num_scanline_pixels": {"dtype": np.int32, "default": 700},
+        "scanline_sector": {"dtype": bool, "default": False},  # native dtype on purpose
```
Comment (wesselvannierop) [inline zea/parameters.py]: `scanline_sector=True` is very similar to the `grid_type="polar"`, maybe we can generalize the `polar` grid instead. I'll have to think a bit about this.

Comment (tristan-deep) [reply zea/parameters.py]: hmm, agreed. Maybe have 4 types of grid types?

Comment (tristan-deep) [reply zea/parameters.py]: "'cartesian', 'polar', 'scanline', 'scanline_polar'"

Comment (wesselvannierop) [reply zea/parameters.py]: We could just have 2 grid-types "cartesian" and "polar" and just have a "scanline: bool"

File: zea/parameters.py (hunk 6 above)
Comment (wesselvannierop) [reply zea/parameters.py]: Yes okay for now.

File: zea/parameters.py (hunk 8 above)
Comment (tristan-deep) [reply zea/parameters.py]: This makes more sense indeed, will update.

File: zea/beamform/pixelgrid.py (hunk 3 above)
Comment (wesselvannierop) [reply zea/beamform/pixelgrid.py:197]: Related to the other discussion so closing this one.

File: zea/parameters.py (hunk 8 above)
Comment (tristan-deep) [reply zea/parameters.py]: Should be refactored, now have `enable_scanline` bool and the familiar grid types. Under the hood refactor was a bit harder, I think it is quite clean like this.

**Comment by tristan-deep** [comment]
Just added a final snippet to the PR description to reproduce the figure.

File: zea/ops/ultrasound.py [hunk 9]
```diff
@@ -238,20 +238,47 @@ def call(self, flat_pfield=None, **kwargs):
...
-        # Add batch dimension if needed
-        if self.with_batch_dim:
-            pfield_expanded = ops.expand_dims(flat_pfield, axis=0)
-        else:
-            pfield_expanded = flat_pfield
+        return {self.output_key: weighted_data}
+
+
+@ops_registry("receive_apodization")
+class ReceiveApodization(Operation):
```
Comment (OisinNolan) [inline zea/ops/ultrasound.py]: Usually when I heard about receive apodization I always associated it with weighing the channels, pre-beamforming, and also seems to be the case when I google it, e.g. https://ieeexplore.ieee.org/document/5442035

but maybe this is one of those ambiguous terms in ultrasound? We could consider going with a name like `PixelApodization` or `AlignedApodization` or something to disambiguate, curious what you think

File: zea/beamform/pixelgrid.py [hunk 10]
```diff
@@ -192,3 +192,108 @@ def polar_pixel_grid(
...
+            np.sin(polar_angles) * np.cos(azimuth_angles),
+            np.sin(polar_angles) * np.sin(azimuth_angles),
+            np.cos(polar_angles),
+        ],
+        axis=-1,
+    )
+
+    # Every scanline column is a ray sampled along ``line_depths``; the two grid
+    # types differ only in each column's origin and direction.
+    if grid_type == "polar":
```
Comment (OisinNolan) [inline zea/beamform/pixelgrid.py:258]: I wonder if we need a polar / cartesian distinction here -- is the general solution just always draw a line from the focal point to the tx_origin and grab the coordinates along that line to beamform?

as an interesting example, check out the pfield from the carotid scan that @pjarosik sent us (it's a linear probe, and we'd typically use a cartesian grid type, but there's some steering at the edges).

https://github.com/user-attachments/assets/54b52ff8-3b0c-4315-bd11-67ebcfeb7c86

File: zea/ops/ultrasound.py (hunk 9 above)
Comment (tristan-deep) [reply zea/ops/ultrasound.py]: You are right, I am changing the renaming and will also try to add a real ReceiveApodization operation.

File: zea/beamform/pixelgrid.py (hunk 10 above)
Comment (tristan-deep) [reply zea/beamform/pixelgrid.py:258]: hmm. That would not be a bad idea. Would you want to give it a go, also given you have some of this data to try it out on?

Comment (tristan-deep) [reply zea/beamform/pixelgrid.py:258]: The idea would be that if `enable_scanline` is set to True, you would not have to specify a grid_type or at least that parameter would not be used? Since it can figure out the grid based on the transmit sequence? That would be kind of neat.

Comment (OisinNolan) [reply zea/beamform/pixelgrid.py:258]: Yeah I'm interesting in looking into this -- now that I think about it though, it may be one for a future PR since it would actually generalize the grid type -- we could end up with grids that are neither polar nor cartesian. So will need to think about a nice way to handle that

will do some testing locally to see if there's a neat solution, but feel free to merge this once it's ready

Comment (tristan-deep) [reply zea/beamform/pixelgrid.py:258]: Okay nice! Maybe something like an "adaptive" grid type would be cool in that case. Would keep old stuff compatible as well. Adaptive in the sense adapt grid to the transmit scheme automatically. 

Do you like the latest changes? Feel free to approve the PR.

**Review by wesselvannierop — APPROVED** [review]
lgtm! I think we might be able to generalize it a bit more in the future given that `f_number`, `pfield` and these new `AlignedApodization` and `ReceiveApodization` all are essentially doing the same thing. Kind of related to https://github.com/tue-bmd/zea/issues/143

**Comment by tristan-deep** [comment]
> lgtm! I think we might be able to generalize it a bit more in the future given that `f_number`, `pfield` and these new `AlignedApodization` and `ReceiveApodization` all are essentially doing the same thing. Kind of related to #143

Indeed they are all very similar and slight variation on how to apply it or on which axis. I tried to at least for now document it a bit in the docstrings. Some cleanup or better integration is probably appreciated in the future.
